llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1385,6 +1387,57 @@ void VPlanTransforms::addActiveLaneMask(
     HeaderMask->replaceAllUsesWith(LaneMask);
 }
 
+/// Replace recipes with their EVL variants.
+static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
+  VPDominatorTree VPDT;
+  VPDT.recalculate(Plan);
+  DenseSet<VPRecipeBase *> ToRemove;
+
+  ReversePostOrderTraversal<VPBlockDeepTraversalWrapper<VPBlockBase *>> RPOT(
+      Plan.getEntry());
+  DenseSet<VPValue *> HeaderMasks = collectAllHeaderMasks(Plan);
+  for (VPBasicBlock *VPBB :
+       reverse(VPBlockUtils::blocksOnly<VPBasicBlock>(RPOT))) {
+    for (VPRecipeBase &R : make_early_inc_range(reverse(*VPBB))) {
+      if (!properlyDominates(EVL.getDefiningRecipe(), &R, VPDT))
```

Comment:
Can this be an assert? We just earlier inserted EVL at a place that should dominate all recipes handled below?

Reply:
It's related to [comment above](https://github.com/llvm/llvm-project/pull/93854#discussion_r1626100405): there are some cases when preheader or postexit may technically contain evl-based recipes, but they do require EVL, but it should be constructed before.

Reply:
will remove it for now as it does assert work for VF=1.